internal/docstore/firedocstore/native_codec_test.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2019 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package firedocstore
+
+import (
+	"context"
+	"io"
+	"net"
+	"testing"
+
+	"cloud.google.com/go/firestore"
+	ts "github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/api/option"
+	pb "google.golang.org/genproto/googleapis/firestore/v1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+// A nativeCodec encodes and decodes structs using the cloud.google.com/go/firestore
+// client. Since that package doesn't export its codec, we have to go behind the
+// scenes and intercept traffic at the gRPC level. We use interceptors to do that. (A
+// mock server would have worked too.)
+
+type nativeCodec struct {
+	client *firestore.Client
+	doc    *pb.Document
+}
+
+func newNativeCodec() (*nativeCodec, error) {
+	// Establish a gRPC server, just so we have a connection to hang the interceptors on.
+	srv := grpc.NewServer()
+	l, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		return nil, err
+	}
+	go func() {
+		if err := srv.Serve(l); err != nil {
+			panic(err) // we should never get an error because we just connect and stop
+		}
+	}()
+	nc := &nativeCodec{}
+
+	conn, err := grpc.Dial(l.Addr().String(),
+		grpc.WithInsecure(),
+		grpc.WithBlock(),
+		grpc.WithUnaryInterceptor(nc.interceptUnary),
+		grpc.WithStreamInterceptor(nc.interceptStream))
+	if err != nil {
+		return nil, err
+	}
+	conn.Close()
+	srv.Stop()
+	nc.client, err = firestore.NewClient(context.Background(), "P", option.WithGRPCConn(conn))
+	if err != nil {
+		return nil, err
+	}
+	return nc, nil
+}
+
+// Intercept all unary (non-streaming) RPCs. The only one we should ever get is a Commit, for
+// the Create call in Encode.
+// If this completes successfully, the encoded *pb.Document will be in c.doc.
+func (c *nativeCodec) interceptUnary(_ context.Context, method string, req, res interface{}, _ *grpc.ClientConn, _ grpc.UnaryInvoker, _ ...grpc.CallOption) error {
+	c.doc = req.(*pb.CommitRequest).Writes[0].GetUpdate()
+	cres := res.(*pb.CommitResponse)
```

Comment:
Where is `cres` used? Why do we need to give it a WriteResult?

Reply:
Got rid of `cres`. We need a WriteResult because the firestore client has a check for it.